### PR TITLE
Substitute hex numbers in word by token

### PR DESCRIPTION
**Pull request by ayush1999**
Currently, any occurrence of a hex number will be replaced by __HEX_NUMBER__ .

### CODE REVIEW

**Comment by ayush1999** [comment]
Results:
Before changes:
```ayush99@ayush99:~/GitHub/bugbug$ python run.py --train --goal=bug
(11173, 344181) (11173,)
(1882, 344181) (1882,)
(1118, 344181) (1118,)
CV Accuracy: 0.92 (+/- 0.02)
```
After changes:
```ayush99@ayush99:~/GitHub/bugbug$ python run.py --train --goal=bug
(11173, 344181) (11173,)
(1882, 344181) (1882,)
(1118, 344181) (1118,)
CV Accuracy: 0.93 (+/- 0.02)
```

**Comment by marco-c** [comment]
It looks like in the `regression` case we are a little worse off. So we should wait as #52 for optional cleanups, or figure out why it's worse (maybe the regex is replacing something it shouldn't?).

**Comment by ayush1999** [comment]
@marco-c Would you like me to create a PR for using optional keywords?

**Comment by marco-c** [comment]
Look at `function_extractors` and you can do something similar (https://github.com/marco-c/bugbug/blob/master/bugbug/bug_features.py#L151). The models will pass the functions, and BugExtractor will call the functions.

**Comment by marco-c** [comment]
Now we can do both this and the other PR, applying the cleanup functions only for the models that are improved.

Before checking which models are improved, please update your bugs DB (just remove the bugs.json file in your data directory and a new one should be downloaded automatically).

**Comment by marco-c** [comment]
Ah, and to check the results, check both the CV accuracy, the first confusion matrix (no confidence level) and the last confusion matrix (0.9 confidence level)

**Comment by ayush1999** [comment]
I'm posting the results here:
Before changes:
(goal=bug)
```
(6528, 240771) (6528,)
(1868, 240771) (1868,)
(653, 240771) (653,)
CV Accuracy: 0.93 (+/- 0.02)
Feature ranking (top 20 features):
1. 'comments_reproduce' (0.027923211455345154)
2. 'comments_pushlog' (0.026178009808063507)
3. 'data_severity=critical' (0.022687608376145363)
4. 'first_comment_str' (0.022687608376145363)
5. 'data_title-fail=True' (0.022687608376145363)
6. 'comments_regression' (0.022687608376145363)
7. 'comments_bug' (0.022687608376145363)
8. 'title___file_reference__' (0.020942408591508865)
9. 'comments_fix' (0.01919720694422722)
10. 'first_comment_regression' (0.01919720694422722)
11. 'comments_need' (0.01919720694422722)
12. 'comments_issue' (0.013961605727672577)
13. 'title_remove' (0.012216405011713505)
14. 'comments_needed' (0.012216405011713505)
15. 'comments_57' (0.012216405011713505)
16. 'first_comment_attachment' (0.012216405011713505)
17. 'comments_currently' (0.012216405011713505)
18. 'first_comment_mozilla' (0.012216405011713505)
19. 'comments_56' (0.012216405011713505)
20. 'first_comment___code_reference_url__' (0.012216405011713505)
No confidence threshold - 653 classified
[[508  55]
 [  4  86]]
                   pre       rec       spe        f1       geo       iba       sup

          1       0.99      0.90      0.96      0.95      0.93      0.86       563
          0       0.61      0.96      0.90      0.74      0.93      0.87        90

avg / total       0.94      0.91      0.95      0.92      0.93      0.86       653


Confidence threshold > 0.6 - 621 classified
[[488  44]
 [  4  85]]
                   pre       rec       spe        f1       geo       iba       sup

          1       0.99      0.92      0.96      0.95      0.94      0.87       532
          0       0.66      0.96      0.92      0.78      0.94      0.88        89

avg / total       0.94      0.92      0.95      0.93      0.94      0.87       621


Confidence threshold > 0.7 - 581 classified
[[465  30]
 [  3  83]]
                   pre       rec       spe        f1       geo       iba       sup

          1       0.99      0.94      0.97      0.97      0.95      0.90       495
          0       0.73      0.97      0.94      0.83      0.95      0.91        86

avg / total       0.96      0.94      0.96      0.95      0.95      0.90       581


Confidence threshold > 0.8 - 525 classified
[[428  19]
 [  0  78]]
                   pre       rec       spe        f1       geo       iba       sup

          1       1.00      0.96      1.00      0.98      0.98      0.95       447
          0       0.80      1.00      0.96      0.89      0.98      0.96        78

avg / total       0.97      0.96      0.99      0.97      0.98      0.95       525


Confidence threshold > 0.9 - 444 classified
[[373   6]
 [  0  65]]
                   pre       rec       spe        f1       geo       iba       sup

          1       1.00      0.98      1.00      0.99      0.99      0.98       379
          0       0.92      1.00      0.98      0.96      0.99      0.99        65

avg / total       0.99      0.99      1.00      0.99      0.99      0.98       444
```
After changes:
```ayush99@ayush99:~/GitHub/bugbug$ python run.py --train --goal=bug
(6528, 205821) (6528,)
(1868, 205821) (1868,)
(653, 205821) (653,)
CV Accuracy: 0.93 (+/- 0.01)
Feature ranking (top 20 features):
1. 'comments_reproduce' (0.027681661769747734)
2. 'comments_pushlog' (0.025951556861400604)
3. 'comments_fix' (0.022491348907351494)
4. 'data_title-fail=True' (0.022491348907351494)
5. 'first_comment_str' (0.020761245861649513)
6. 'comments_regression' (0.020761245861649513)
7. 'data_severity=critical' (0.020761245861649513)
8. 'first_comment_regression' (0.019031140953302383)
9. 'comments_bug' (0.017301037907600403)
10. 'title___file_reference__' (0.015570933930575848)
11. 'first_comment_mozilla' (0.013840830884873867)
12. 'comments_need' (0.013840830884873867)
13. 'comments_issue' (0.013840830884873867)
14. 'comments_approved' (0.013840830884873867)
15. 'comments_10' (0.013840830884873867)
16. 'comments_remove' (0.012110726907849312)
17. 'title_remove' (0.012110726907849312)
18. 'comments_currently' (0.012110726907849312)
19. 'data_keywords-crash=True' (0.012110726907849312)
20. 'comments_nightly' (0.012110726907849312)
No confidence threshold - 653 classified
[[507  56]
 [  5  85]]
                   pre       rec       spe        f1       geo       iba       sup

          1       0.99      0.90      0.94      0.94      0.92      0.85       563
          0       0.60      0.94      0.90      0.74      0.92      0.85        90

avg / total       0.94      0.91      0.94      0.91      0.92      0.85       653


Confidence threshold > 0.6 - 621 classified
[[489  44]
 [  3  85]]
                   pre       rec       spe        f1       geo       iba       sup

          1       0.99      0.92      0.97      0.95      0.94      0.88       533
          0       0.66      0.97      0.92      0.78      0.94      0.89        88

avg / total       0.95      0.92      0.96      0.93      0.94      0.88       621


Confidence threshold > 0.7 - 584 classified
[[468  33]
 [  3  80]]
                   pre       rec       spe        f1       geo       iba       sup

          1       0.99      0.93      0.96      0.96      0.95      0.90       501
          0       0.71      0.96      0.93      0.82      0.95      0.90        83

avg / total       0.95      0.94      0.96      0.94      0.95      0.90       584


Confidence threshold > 0.8 - 525 classified
[[430  20]
 [  0  75]]
                   pre       rec       spe        f1       geo       iba       sup

          1       1.00      0.96      1.00      0.98      0.98      0.95       450
          0       0.79      1.00      0.96      0.88      0.98      0.96        75

avg / total       0.97      0.96      0.99      0.96      0.98      0.95       525


Confidence threshold > 0.9 - 443 classified
[[370   7]
 [  0  66]]
                   pre       rec       spe        f1       geo       iba       sup

          1       1.00      0.98      1.00      0.99      0.99      0.98       377
          0       0.90      1.00      0.98      0.95      0.99      0.98        66

avg / total       0.99      0.98      1.00      0.98      0.99      0.98       443
```

**Comment by ayush1999** [comment]
goal = qaneeded
Before:
```ayush99@ayush99:~/GitHub/bugbug$ python run.py --train --goal=qaneeded
(59876, 744318) (59876,)
(66, 744318) (66,)
(5988, 744318) (5988,)
CV Accuracy: 0.74 (+/- 0.16)
Feature ranking (top 20 features):
1. 'comments_reply' (0.1294642835855484)
2. 'comments_10' (0.0669642835855484)
3. 'comments_sure' (0.0625)
4. 'comments_happens' (0.0535714291036129)
5. 'comments_don' (0.0535714291036129)
6. 'comments_str' (0.0446428582072258)
7. 'comments_just' (0.0446428582072258)
8. 'comments_like' (0.0401785708963871)
9. 'comments_attachment' (0.03125)
10. 'comments_firefox' (0.03125)
11. 'comments_profile' (0.0223214291036129)
12. 'comments_open' (0.0223214291036129)
13. 'comments_related' (0.0223214291036129)
14. 'comments_did' (0.0223214291036129)
15. 'comments_current' (0.0223214291036129)
16. 'comments_build' (0.01785714365541935)
17. 'comments_time' (0.01785714365541935)
18. 'comments_reproduce' (0.01785714365541935)
19. 'comments_58' (0.013392857275903225)
20. 'comments_hi' (0.013392857275903225)
No confidence threshold - 5988 classified
[[   3    0]
 [ 930 5055]]
                   pre       rec       spe        f1       geo       iba       sup

          1       0.00      1.00      0.84      0.01      0.92      0.86         3
          0       1.00      0.84      1.00      0.92      0.92      0.83      5985

avg / total       1.00      0.84      1.00      0.92      0.92      0.83      5988


Confidence threshold > 0.6 - 5610 classified
[[   3    0]
 [ 751 4856]]
                   pre       rec       spe        f1       geo       iba       sup

          1       0.00      1.00      0.87      0.01      0.93      0.88         3
          0       1.00      0.87      1.00      0.93      0.93      0.85      5607

avg / total       1.00      0.87      1.00      0.93      0.93      0.85      5610


Confidence threshold > 0.7 - 5253 classified
[[   3    0]
 [ 572 4678]]
                   pre       rec       spe        f1       geo       iba       sup

          1       0.01      1.00      0.89      0.01      0.94      0.90         3
          0       1.00      0.89      1.00      0.94      0.94      0.88      5250

avg / total       1.00      0.89      1.00      0.94      0.94      0.88      5253


Confidence threshold > 0.8 - 4802 classified
[[   3    0]
 [ 415 4384]]
                   pre       rec       spe        f1       geo       iba       sup

          1       0.01      1.00      0.91      0.01      0.96      0.92         3
          0       1.00      0.91      1.00      0.95      0.96      0.91      4799

avg / total       1.00      0.91      1.00      0.95      0.96      0.91      4802


Confidence threshold > 0.9 - 4038 classified
[[   2    0]
 [ 222 3814]]
                   pre       rec       spe        f1       geo       iba       sup

          1       0.01      1.00      0.94      0.02      0.97      0.95         2
          0       1.00      0.94      1.00      0.97      0.97      0.94      4036

avg / total       1.00      0.95      1.00      0.97      0.97      0.94      4038
```
After:
```ayush99@ayush99:~/GitHub/bugbug$ python run.py --train --goal=qaneeded
(59876, 627506) (59876,)
(66, 627506) (66,)
(5988, 627506) (5988,)
CV Accuracy: 0.74 (+/- 0.16)
Feature ranking (top 20 features):
1. 'comments_reply' (0.1294642835855484)
2. 'comments_10' (0.0669642835855484)
3. 'comments_sure' (0.0625)
4. 'comments_happens' (0.0535714291036129)
5. 'comments_don' (0.0535714291036129)
6. 'comments_just' (0.0446428582072258)
7. 'comments_str' (0.0446428582072258)
8. 'comments_like' (0.0401785708963871)
9. 'comments_firefox' (0.03125)
10. 'comments_attachment' (0.03125)
11. 'comments_related' (0.0223214291036129)
12. 'comments_profile' (0.0223214291036129)
13. 'comments_current' (0.0223214291036129)
14. 'comments_open' (0.0223214291036129)
15. 'comments_did' (0.0223214291036129)
16. 'comments_reproduce' (0.01785714365541935)
17. 'comments_time' (0.01785714365541935)
18. 'comments_build' (0.01785714365541935)
19. 'comments_hi' (0.013392857275903225)
20. 'comments_58' (0.013392857275903225)
No confidence threshold - 5988 classified
[[   3    0]
 [ 928 5057]]
                   pre       rec       spe        f1       geo       iba       sup

          1       0.00      1.00      0.84      0.01      0.92      0.86         3
          0       1.00      0.84      1.00      0.92      0.92      0.83      5985

avg / total       1.00      0.85      1.00      0.92      0.92      0.83      5988


Confidence threshold > 0.6 - 5610 classified
[[   3    0]
 [ 751 4856]]
                   pre       rec       spe        f1       geo       iba       sup

          1       0.00      1.00      0.87      0.01      0.93      0.88         3
          0       1.00      0.87      1.00      0.93      0.93      0.85      5607

avg / total       1.00      0.87      1.00      0.93      0.93      0.85      5610


Confidence threshold > 0.7 - 5253 classified
[[   3    0]
 [ 572 4678]]
                   pre       rec       spe        f1       geo       iba       sup

          1       0.01      1.00      0.89      0.01      0.94      0.90         3
          0       1.00      0.89      1.00      0.94      0.94      0.88      5250

avg / total       1.00      0.89      1.00      0.94      0.94      0.88      5253


Confidence threshold > 0.8 - 4800 classified
[[   3    0]
 [ 413 4384]]
                   pre       rec       spe        f1       geo       iba       sup

          1       0.01      1.00      0.91      0.01      0.96      0.92         3
          0       1.00      0.91      1.00      0.96      0.96      0.91      4797

avg / total       1.00      0.91      1.00      0.95      0.96      0.91      4800


Confidence threshold > 0.9 - 4039 classified
[[   2    0]
 [ 222 3815]]
                   pre       rec       spe        f1       geo       iba       sup

          1       0.01      1.00      0.95      0.02      0.97      0.95         2
          0       1.00      0.95      1.00      0.97      0.97      0.94      4037

avg / total       1.00      0.95      1.00      0.97      0.97      0.94      4039
```

**Comment by ayush1999** [comment]
@marco-c I think for now we should add this option only for the bug and qaneeded case. Any thoughts?

**Comment by marco-c** [comment]
@ayush1999 let's just add the function for now, and then we can evaluate later on whether to use it or not and for which models.

File: bugbug/bug_features.py [hunk 1]
```diff
@@ -132,6 +132,10 @@ def cleanup_fileref(text):
     return re.sub(r'\w+\.py\b|\w+\.json\b|\w+\.js\b|\w+\.jsm\b|\w+\.html\b|\w+\.css\b|\w+\.c\b|\w+\.cpp\b|\w+\.h\b', '__FILE_REFERENCE__', text)
 
 
+def cleanup_hex(text):
+    return re.sub(r'0[xX][0-9a-fA-F]+', '__HEX_NUMBER__', text)
```
Comment (marco-c) [inline bugbug/bug_features.py]: Let's match only if the word ends with the hex number.

**Comment by ayush1999** [comment]
Changes made: Removed all calls for `cleanup_hex`. The function now replaces all words which end with a hex number by the __hex_number__ token.

File: bugbug/bug_features.py [hunk 2]
```diff
@@ -134,6 +134,10 @@ def cleanup_fileref(text):
     return re.sub(r'\w+\.py\b|\w+\.json\b|\w+\.js\b|\w+\.jsm\b|\w+\.html\b|\w+\.css\b|\w+\.c\b|\w+\.cpp\b|\w+\.h\b', '__FILE_REFERENCE__', text)
 
 
+def cleanup_hex(text):
+    return re.sub(r'[^\s]*0[xX][0-9a-fA-F]+', '__HEX_NUMBER__', text)
```
Comment (marco-c) [inline bugbug/bug_features.py]: Same here, you can use `\b` to define a word boundary (and it should be both before and after the number).

File: tests/test_cleanup_functions.py [hunk 3]
```diff
@@ -25,6 +25,14 @@ def test_cleanup_fileref():
         assert bug_features.cleanup_fileref(orig_text) == cleaned_text
 
 
+def test_cleanup_hex():
+    tests = [
+        ('0 scdetour.dll scdetour.dll@0x2dd77', '0 scdetour.dll __HEX_NUMBER__'),
```
Comment (marco-c) [inline tests/test_cleanup_functions.py]: There seems to be something wrong in the result of the test, it should be `0 scdetour.dll scdetour.dll@__HEX_NUMBER`.

Also add a word which is just an hex number in the test.

**Comment by ayush1999** [comment]
@marco-c Done.

**Pull request merged**